lib/src/css_mirroring/directives_processor.dart

```diff
@@ -0,0 +1,45 @@
+part of scissors.src.css_mirroring.bidi_css_generator;
+
+class DirectionIndependentDirectivesProcessor extends NodeProcessor {
+  DirectionIndependentDirectivesProcessor();
+
+  process(EditConfiguration _editConfig, TopLevelsPair _topLevelsPair,
+      int _iTopLevel) {
+    if (_editConfig.mode != _RetentionMode.keepBidiNeutral) {
+      _removeRuleSet(_editConfig.transaction,
+          _getUsedTopLevels(_editConfig.mode, _topLevelsPair), _iTopLevel);
+    }
+  }
+}
+
+class DirectionDependentDirectivesProcessor extends NodeProcessor {
+  /// All removable declarations of ruleset are removed and if all declarations
+  /// in rulesets have to be removed, it removes ruleset itself.
+  /// Also if all rulesets have to be removed, it removes the directive.
+  process(EditConfiguration _editConfig, TopLevelsPair _topLevelsPair,
+      int _iTopLevel) {
+    final _originalRuleSets =
+        _topLevelsPair._originalTopLevels[_iTopLevel].rulesets;
+    final _flippedRuleSets =
+        _topLevelsPair._flippedTopLevels[_iTopLevel].rulesets;
+    TopLevelsPair ruleSetsPair =
+        new TopLevelsPair(_originalRuleSets, _flippedRuleSets);
+    final _usedDirective =
+        _getUsedTopLevels(_editConfig.mode, _topLevelsPair)[_iTopLevel];
+    final _removableRuleSets = new _PendingRemovals(_editConfig.transaction);
+
+    ruleSetsPair.forEach((original, flipped, iDecl) {
+      new DeclarationProcessor(
+              _removableRuleSets, _usedDirective.span.end.offset)
+          .process(_editConfig, ruleSetsPair, iDecl);
+    });
+
+    if (_removableRuleSets._removalStartEndLocations.length ==
```

Comment:
using a private field of another class is a code smell: just expose a `bool get length`